server/lib/src/idm/identityverification.rs

```diff
@@ -0,0 +1,611 @@
+            .expect("Failed to impersonate identity");
+
+        let res = idms_prox_read.handle_identify_user_start(&IdentifyUserStartEvent::new(
+            valid_user_a_uuid,
+            ident.clone(),
+        ));
+
+        assert!(matches!(
+            res,
+            Ok(IdentifyUserResponse::IdentityVerificationAvailable)
+        ));
+    }
+
+    #[idm_test] // actually this is somewhat a duplicate of `test_full_identification_flow` inside the testkit, with the exception that this
+                //tests ONLY the totp code correctness and not the flow correctness. To test the correctness it obviously needs to also
+                // enforce some flow checks, but this is not the primary scope of this test
+    async fn test_code_correctness(idms: &IdmServer, _idms_delayed: &IdmServerDelayed) {
+        let ct = duration_from_epoch_now();
+        let mut idms_prox_write = idms.proxy_write(ct).await;
+        let user_a_uuid = Uuid::new_v4();
+        let user_b_uuid = Uuid::new_v4();
+        let e1 = create_valid_user_account(user_a_uuid);
+        let e2 = create_valid_user_account(user_b_uuid);
+        let ce = CreateEvent::new_internal(vec![e1, e2]);
+
+        assert!(idms_prox_write.qs_write.create(&ce).is_ok());
+        assert!(idms_prox_write.commit().is_ok());
+
+        let mut idms_prox_read = idms.proxy_read().await;
+
+        let ident_a = idms_prox_read
+            .qs_read
+            .internal_search_uuid(user_a_uuid)
+            .map(Identity::from_impersonate_entry_readonly)
+            .expect("Failed to impersonate identity");
+
+        let ident_b = idms_prox_read
+            .qs_read
+            .internal_search_uuid(user_b_uuid)
+            .map(Identity::from_impersonate_entry_readonly)
+            .expect("Failed to impersonate identity");
+
+        // If a has the lowest uuid then they are going to receive the code from b
+        if user_a_uuid < user_b_uuid {
+            // from here we get the code from user b and submit it as user a
+            let res_b = idms_prox_read.handle_identify_user_start(&IdentifyUserStartEvent::new(
+                user_a_uuid,
+                ident_b.clone(),
+            ));
+
+            let Ok(IdentifyUserResponse::ProvideCode { totp, .. }) = res_b else {
+                return assert!(false);
+            };
+
+            let res_a_wrong = idms_prox_read.handle_identify_user_submit_code(
+                &IdentifyUserSubmitCodeEvent::new(user_b_uuid, ident_a.clone(), totp + 1),
+            );
+
+            assert!(matches!(res_a_wrong, Ok(IdentifyUserResponse::CodeFailure)));
+
+            let res_a_correct = idms_prox_read.handle_identify_user_submit_code(
+                &IdentifyUserSubmitCodeEvent::new(user_b_uuid, ident_a.clone(), totp),
+            );
+
+            assert!(matches!(
+                res_a_correct,
+                Ok(IdentifyUserResponse::ProvideCode { .. })
+            ));
+
```

Comment:
Add a test that asserts that the totps on both sides are never the same. The totp I give you shouldn't be the same as the totp you give me. I think you've missed this in your cryptographic code, I can't see where you invert the secret based on uuid order.

Reply:
Basically from the document we wrote back in march we ruled that each user secret would be composed by the shared ecdh secret and the user own uuid, so `user a` will have `totp_secret_a = shared_ecdh_secret + user_a_uuid` and `user b` will have `totp_secret_b = shared_ecdh_secret + user_b_uuid`

In `get_self_totp_secret` I have
> let Some(self_uuid) = ident.get_uuid() else {
            return Err(OperationError::NotAuthenticated)
        };
        shared_key.extend_from_slice(self_uuid.as_bytes());


while in `get_other_user_totp_secret`
>  shared_key.extend_from_slice(target.as_bytes());

Reply:
Still good to have a test case proving that the totps aren't the same :)